homeassistant/components/mikrotik/__init__.py

```diff
@@ -1 +1,274 @@
+                self._host)
+            self._connected = False
+            return
+        self.hass.data[DOMAIN][self._host]['info'] = data[0]
+
+    async def update_device_tracker(self, method=None):
+        """Update device_tracker from Mikrotik API."""
+        self.hass.data[DOMAIN][self._host][DEVICE_TRACKER] = {}
+        if method is None:
+            return
+        _LOGGER.debug(
+            "[%s] Updating Mikrotik device_tracker using %s.",
+            self._host, method)
+
+        data = self.get_api(MIKROTIK_SERVICES[method])
+        if data is None:
+            self.update_info()
+            return
+
+        arp = self.get_api(MIKROTIK_SERVICES[ARP])
+        for device in arp:
+            if 'mac-address' in device and device['invalid'] is False:
+                mac = device['mac-address']
+                self.hass.data[DOMAIN][ARP][mac] = device
+
+        for device in data:
+            mac = device['mac-address']
+            if method == DHCP:
+                if 'active-address' not in device:
+                    continue
+                self.hass.data[DOMAIN][DHCP][mac] = data
+                if (self._arp_ping and mac in arp):
+                    interface = arp[mac]['interface']
+                    if not self.arp_ping(mac, interface):
+                        continue
+
+            attributes = {}
+            for attrib in ATTR_DEVICE_TRACKER:
+                if attrib in device:
+                    attributes[slugify(attrib)] = device[attrib]
+            attributes['source_type'] = 'router'
+            attributes['scanner_type'] = method
+            attributes['scanner_host'] = self._host
+            attributes['scanner_host_name'] = self._host_name
+
+            if mac in self.hass.data[DOMAIN][ARP]:
+                attributes['ip_address'] = self.hass.data[DOMAIN][
+                    ARP][mac]['address']
+
+            if mac in self.hass.data[DOMAIN][DHCP]:
+                attributes['host_name'] = self.hass.data[DOMAIN][
+                    DHCP][mac]['host-name']
+
+            self.hass.data[DOMAIN][self._host][
+                DEVICE_TRACKER][mac] = attributes
+
+    def get_api(self, cmd, params=None):
+        """Retrieve data from Mikrotik API."""
+        if not self._client or not self._connected:
+            if not self.connect_to_device():
+                return None
+        try:
+            if params:
+                response = self._client(cmd=cmd, **params)
+            else:
+                response = self._client(cmd=cmd)
+        except (librouteros.exceptions.TrapError,
+                librouteros.exceptions.MultiTrapError,
+                librouteros.exceptions.ConnectionError) as api_error:
+            _LOGGER.error(
+                "Failed to retrieve data. "
+                "%s cmd=[%s] Error: %s", self._host, cmd, api_error)
+            self._connected = False
+            return None
+        return response
+
+
+class MikrotikClient:
```

Comment:
I suggest we remove this class and just extend the `MikrotikAPI` class as needed. Is there some separation of concerns that is appropriate here do you think?

Reply:
I agree, as I was writing that class I did think it was unnecessary. I will make some changes and re-commit them. Thank you for all the suggestions.